packages/core/src/locator.ts

```diff
@@ -18,73 +25,121 @@ function baseInstantiator<Context>(opts: {
+
+  constructor(private parent?: OverrideInterceptor<C>) {}
+
+  get: InterceptorGet<C> = i => f => {
+    if (this.overrides.has(f)) {
+      f = this.overrides.get(f) as any;
+    } else if (this.parent && this.parent.has(f)) {
+      return this.parent.get(i)(f) as any;
     }
-    return instantiator(f);
+
+    return i(f);
   };
+
+  has<T>(f: ConstructorOrFactory<C, T>): boolean {
+    let found = this.overrides.has(f);
+    if (!found && this.parent) {
+      found = this.parent.has(f);
+    }
+    return found;
+  }
+
+  override<T>(f: ConstructorOrFactory<C, T>, g: ConstructorOrFactory<C, T>) {
+    this.overrides.set(f, g);
+  }
+
+  clear() {
+    this.overrides.clear();
+  }
+
+  inherit(): Interceptor<C> {
+    const i = new OverrideInterceptor<C>(this);
+    return i;
+  }
 }
 
 export class Locator<Context> {
-  private instances: Map<Function, any> = new Map();
-  private overrides: Map<Function, Function> = new Map();
-
-  public get: <T>(f: ConstructorOrFactory<Context, T>) => T;
+  private interceptors: Interceptor<Context>[] = [];
+  public baseGet: <T>(f: ConstructorOrFactory<Context, T>) => T;
 
   constructor(
     locatorCtx: Context,
     private isClass: (v: ConstructorOrFactory<Context, any>) => boolean,
-    options: {
-      overrides?: Map<Function, Function>;
-      isTransient?: boolean;
-    },
   ) {
-    if (options.overrides != null) this.overrides = options.overrides;
-    this.get = baseInstantiator({ isClass, locatorCtx });
-    if (!options.isTransient) this.addInterceptor(cachingInterceptor(this.instances));
-    this.addInterceptor(overrideInterceptor(this.overrides));
+    this.baseGet = baseInstantiator({ isClass, locatorCtx });
   }
 
   public addInterceptor(ic: Interceptor<Context>) {
-    this.get = ic(this.get);
+    this.interceptors.push(ic);
   }
 
+  get = <T>(f: ConstructorOrFactory<Context, T>) => {
+    const ics = this.interceptors.filter(i => i.get) as Interceptor<Context>[];
+    const reducedFn = ics.reduce((acc, ic) => ic.get!(acc), this.baseGet);
+    return reducedFn(f);
+  };
+
   has<T>(f: ConstructorOrFactory<Context, T>): boolean {
-    if (this.instances.has(f)) return true;
-    let override = this.overrides.get(f);
-    if (override) return this.instances.has(override);
-    return false;
+    const ics = this.interceptors.filter(i => i.has);
+    const found = ics.some(ic => ic.has!(f));
+    return found;
   }
 
   override<T>(f: ConstructorOrFactory<Context, T>, g: ConstructorOrFactory<Context, T>) {
-    if (this.instances.has(f)) {
-      console.warn(
-        `Warning: by overriding ${f.name}, you will be shadowing an already instantiated class.`,
-      );
-    }
-    this.overrides.set(f, g);
+    const ics = this.interceptors.filter(i => i.override);
+    ics.forEach(i => i.override!(f, g));
   }
 
   withNewContext(ctx: Context) {
-    return new Locator(ctx, this.isClass, { overrides: this.overrides });
+    const loc = new Locator(ctx, this.isClass);
```

Comment:
it seems some of the interceptors will be duplicates because they're already added in the constructor, right?

you might need a 3rd internal argument which if passed will initialize a locator without any interceptors

Reply:
No, I removed adding interceptors in the constructor https://github.com/hfour/h4bff/pull/77/files#diff-8049e69ed366eddd2272492676109691R104. Do you still think we should make the initialization optional?

Reply:
I think externally we should not complicate the interface of interceptors, internally how we handle the overrides etc is less important and we can figure out a more sensible way later.